### DISCUSSION TITLE
Support for sliding window / moving averages

### DISCUSSION THREAD

**Discussion by mationai** [questdb/questdb#1162] [Ideas]
A Sliding window or Moving average is pretty important in many use cases. It would be nice if questdb has this built in.

For reference, here's clickhouse's:
https://clickhouse.tech/docs/en/sql-reference/aggregate-functions/reference/grouparraymovingavg/

**Comment by newskooler** [comment]
To add to this discussion - I believe "the python idea" may be relevant here. 
@bluestreak01 can comment more here; correct me if I am wrong but my understanding goes like this:
Option 1 (with no python support in QDB) then the team needs to add support for all kind of this such as:
- simple moving average
- exponential moving everage
- (any other) moving average - the concept of rolling in general (first / last / sum, etc)

Option 2 - with python support over QDB data running natively (and by extention much quickly - which is the point of this); then the QDB team don't need to add SQL support for the above mentioned, but just the python one.

Is my understanding correct @bluestreak01  or no?

**Reply by doronaviguy** [reply]
+1

**Reply by doronaviguy** [reply]
In general usually time series queries are usually based on moving avg , median, percentiles , all sort of aggregation functions that other DBs have out of the box ,

the performance are quite good in questdb, 
but the lack of aggregation functions creates a real problem for a lot of use cases

**Comment by marregui** [comment]
Hi,

I am writing this here as a spec on which to base discussions toward the design of this feature. 

My understanding of window functions, these are functions that are applied to a set of rows around the current row (comparable to aggregation functions but rows are not grouped into a single row at the end).

### Syntax

```ebnf
function_name ([expr [, expr ... ]])
              OVER ( [ window_definition ] )
```

The presence of OVER makes the function a window function. 
An empty OVER() defines a window with all the rows in the result set.

**function_name**: general-purpose window or aggregate function one of **[**
```sql
    row_number, first_value, last_value, nth_value, lag, lead, rank, dense_rank, (window functions)
    avg, count, count_distinct, first, haversine_dist_deg, ksum, last, max, min, nsum, sum (aggregate functions)
    mean, stddev, variance, percentile (aggregate functions extra)
```
**]**
**expr**:  column reference, function or constant, or * for count function.
**window_definition**: 
```ebnf
   [ PARTITION BY expr [, ...] ]
   [ ORDER BY expr [ ASC | DESC ]]
   [ frame_spec ]
```
**frame_spec**:
```ebnf
  { RANGE | ROWS } BETWEEN frame_start AND frame_end 
```
**frame_start**, **frame_end**:  
```
   UNBOUNDED PRECEDING
   offset PRECEDING 
   CURRENT ROW
   offset FOLLOWING
   UNBOUNDED FOLLOWING
```

Default frame spec: 
   - RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW. 
   - When frame_end is missing it defaults to CURRENT ROW.
   - frame_start cannot be FOLLOWING or UNBOUNDED FOLLOWING
   - frame_end cannot be PRECEDING or UNBOUNDED PRECEDING.
   - In ROWS mode:
     + CURRENT_ROW is the current row.
     + **offset**: frame start/end is that many rows before/after the current row. 
   - In RANGE mode:
     + If frame_start is CURRENT ROW the frame starts with the current row’s first mate (the row the window’s ORDER BY expr deems as equal to current row).
     + If frame_end is CURRENT ROW the frame ends with the current’s row last mate.
     + **offset**: requires exactly one ORDER BY column in the window definition. 
         contains the rows with ordering column value <= offset minus he current row’s ordering column value (PRECEDING). 
         contains the rows with ordering column value <= offset plus the current row’s ordering column value (FOLLOWING).

**Comment by bsmth** [comment]
Relates to https://github.com/questdb/questdb/issues/1268

**Comment by marregui** [comment]
Relates to https://github.com/questdb/questdb/issues/258

**Comment by bsmth** [comment]
We've created an issue to track the progress of this feature -> https://github.com/questdb/questdb/issues/1351

**Comment by javier** [comment]
related to https://github.com/questdb/questdb/issues/2995

**Comment by ideoma** [comment]
Window functions and moving averages are supported now.
